### PR TITLE
Add `Closure` support to NavigationItem

### PR DESCRIPTION
**Pull request by Johannes-Werbrouck**
- [x] Changes have been thoroughly tested to not break existing functionality.
- [x] New functionality has been documented or existing documentation has been updated to reflect changes.
- [x] Visual changes are explained in the PR description using a screenshot/recording of before and after.

### CODE REVIEW

**Comment by Johannes-Werbrouck** [comment]
**How it works**
Added lazy evaluation on all methods of the `NavigationItem` class. This enables users to enter a callback instead of a string  for things like label, icon, url, badge, ... when using the NavigationBuilder. This way, strings that cannot be parsed yet when building the navigation can be parsed later on, for example when `auth()->user()` is needed.

The `$badgeCallback` takes priority over the `$badge` property, for now. It has been set up much like the `$isActiveWhen` callback. 

**Documentation**
Added some documentation under 'advanced navigation customization'

**Visual changes**
There should be no visual changes.

**Comment by zepfietje** [comment]
@Johannes-Werbrouck, could you update this PR to use `$this->evaluate()` as used everywhere else in the Filament codebase?

**Comment by Johannes-Werbrouck** [comment]
> @Johannes-Werbrouck, could you update this PR to use `$this->evaluate()` as used everywhere else in the Filament codebase?

Done! It now uses the EvaluatesClosures trait.

File: packages/admin/src/Navigation/NavigationItem.php [hunk 1]
```diff
@@ -3,28 +3,31 @@
 namespace Filament\Navigation;
 
 use Closure;
+use Filament\Support\Concerns\EvaluatesClosures;
 
 class NavigationItem
 {
-    protected ?string $group = null;
+    use EvaluatesClosures;
```
Comment (zepfietje) [inline packages/admin/src/Navigation/NavigationItem.php]: @danharrin, should it extend our `Component` class instead of manually using the trait?

**Comment by danharrin** [comment]
Thanks

File: packages/admin/docs/05-navigation.md [hunk 2]
```diff
@@ -188,6 +188,12 @@ Filament::navigation(function (NavigationBuilder $builder): NavigationBuilder {
 });
 ```
 
+The NavigationItem class uses lazy evaluation, so Closures can be used instead of strings for things like `url()`, `icon()`, `badge()`, ... . This enables you to vary the badge color based on a model method, for example:
```
Comment (zepfietje) [inline packages/admin/docs/05-navigation.md:191]: @danharrin did you still rewrite this part of the docs? Don't like how it reads and how it's formatted.

Comment (danharrin) [reply packages/admin/docs/05-navigation.md:191]: I removed it for now

Comment (zepfietje) [reply packages/admin/docs/05-navigation.md:191]: Perfect.

**Pull request merged**